File: java/leetcode/src/Task210.java
import org.junit.jupiter.api.Test;

import java.util.*;

public class Task210 {
    //bfs
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        int[] inCnt = new int[numCourses];
        List<Integer>[] adjList = new List[numCourses];
        for (int i = 0; i < adjList.length; i++) {
            adjList[i] = new ArrayList<>();
        }
        for (int[] prerequisite : prerequisites) {
            ++inCnt[prerequisite[0]];
            adjList[prerequisite[1]].add(prerequisite[0]);
        }

        int[] res = new int[numCourses];
        int idx = 0;
        Queue<Integer> queue = new LinkedList<>();

        for (int i = 0; i < inCnt.length; i++) {
            if (inCnt[i] == 0) {
                queue.offer(i);
            }
        }

        while (!queue.isEmpty()) {
            Integer cur = queue.poll();
            res[idx++] = cur;
            adjList[cur].forEach(val -> {
                --inCnt[val];
                if (inCnt[val] == 0) {
                    queue.offer(val);
                }
            });

        }

        if (idx != numCourses) {
            return new int[0];
        }

        return res;

    }

    //dfs
    enum Color {
        WHITE,
        GRAY,
        BLACK;
    }

    public int[] findOrder2(int numCourses, int[][] prerequisites) {
        List<Integer>[] adjList = new ArrayList[numCourses];
        Color[] colors = new Color[numCourses];
        for (int i = 0; i < numCourses; i++) {
            adjList[i] = new ArrayList<>();
            colors[i] = Color.WHITE;
        }
        for (int[] prerequisite : prerequisites) {
            adjList[prerequisite[0]].add(prerequisite[1]);
        }

        List<Integer> res = new ArrayList<>(numCourses);

        for (int i = 0; i < numCourses; i++) {
            if (!dfs(adjList, i, colors, res)) {
                return new int[0];
            }
        }

        return res.stream().mapToInt(val -> val).toArray();

    }

    private boolean dfs(List<Integer>[] adjList, int val, Color[] colors, List<Integer> res) {
        //terminator
        if (colors[val] == Color.GRAY) {
            return false;
        }
        if (colors[val] == Color.BLACK) {
            return true;
        }
        //process
        //change color into gray
        colors[val] = Color.GRAY;

        //drill down
        for (Integer integer : adjList[val]) {
            if (!dfs(adjList, integer, colors, res)) {
                return false;
            }
        }
        colors[val] = Color.BLACK;
        res.add(val);


        return true;
    }

    @Test
    void test() {
        int[][] prerequisites = new int[][]{{1, 0}, {2, 0}, {3, 1}, {3, 2}};
        int numCourses = 4;
        System.out.println(findOrder2(numCourses, prerequisites));
    }
}

File: py/Task210.py
from typing import List
from collections import defaultdict
from queue import SimpleQueue


class Solution:
    # bfs
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        cnts = [0] * numCourses
        adjList = defaultdict(list)

        for a, b in prerequisites:
            cnts[a] += 1
            adjList[b].append(a)

        q = SimpleQueue()
        res = []
        for idx, cnt in enumerate(cnts):
            if cnt == 0:
                q.put(idx)
                res.append(idx)

        while not q.empty():
            for i in adjList[q.get()]:
                cnts[i] -= 1
                if cnts[i] == 0:
                    q.put(i)
                    res.append(i)

        return res if len(res) == numCourses else []

    # dfs
    def findOrder2(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        colors = [0] * numCourses
        adjList = defaultdict(list)

        res = []

        for a, b in prerequisites:
            adjList[a].append(b)

        def dfs(i: int) -> bool:
            if colors[i] == 1:
                return False
            if colors[i] == 2:
                return True

            colors[i] = 1
            for need in adjList[i]:
                if not dfs(need):
                    return False
            colors[i] = 2
            res.append(i)
            return True

        for i in range(numCourses):
            if not dfs(i):
                return []

        return res


numCourses = 4
prerequisites = [[1, 0], [2, 0], [3, 1], [3, 2]]
obj = Solution()
print(obj.findOrder2(numCourses, prerequisites))
